server/src/PlayerQueue.ts

```diff
@@ -1,39 +1,100 @@
+import { ToClientEvents, ToServerEvents } from "common/messages/sceneWaitingRoom";
+import { Socket } from "socket.io";
+import { broadcastRemainingPlayers, broadcastToSceneGameArena} from "..";
+
+type SocketQueue = Socket<ToServerEvents, ToClientEvents>;
+
 export class PlayerQueue {
-    private playerCounter: number;
+    private queue: Array<SocketQueue>;
+    private connectionsToRemove: Array<SocketQueue>;
 
     constructor() {
-        this.playerCounter = 0;
+        this.queue = [];
+        this.connectionsToRemove = [];
+    }
+
+    public initSocketListeners(socket: SocketQueue) {
+        socket.on("joinQueue", () => {
+            this.clearClosedConnections();
+            this.addToQueue(socket);
+
+            broadcastRemainingPlayers(this.getRemainingPlayers());
+
+            // 4 valid connections are found. Start the game.
+            if (this.getRemainingPlayers() === 0) {
+                for (let playerIndex = 0; playerIndex < 4; playerIndex++) {
+                    this.queue[playerIndex].emit("initPlayer", playerIndex as 0 | 1 | 2 | 3);
+                }
+                broadcastToSceneGameArena();
+            }
+        });
+
+        socket.on("requestRemainingPlayers", () => {
+            socket.emit("updateRemainingPlayers", this.getRemainingPlayers());
+        });
+
+        socket.on("leaveQueue", () => {
```

Comment:
Maybe we should use "disconnect" event here? It's triggered by socket io whenever client socket disconnects. From what I tried it works pretty well. (I don't see leaveQueue emitted anywhere)

Reply:
good catch, that was an old event I forgot to remove.